aws/resource_aws_pinpoint_app.go

```diff
@@ -0,0 +1,435 @@
+	//}
+
+	if d.HasChange("campaign_hook") {
+		appSettings.CampaignHook = expandCampaignHook(d)
+	}
+
+	if d.HasChange("limits") {
+		appSettings.Limits = expandCampaignLimits(d.Get("limits").(*schema.Set))
+	}
+
+	if d.HasChange("quiet_time") {
+		appSettings.QuietTime = expandQuietTime(d.Get("quiet_time").(*schema.Set))
+	}
+
+	req := pinpoint.UpdateApplicationSettingsInput{
+		ApplicationId:                   aws.String(d.Id()),
+		WriteApplicationSettingsRequest: appSettings,
+	}
+
+	_, err := conn.UpdateApplicationSettings(&req)
+	if err != nil {
+		return err
+	}
+
+	return resourceAwsPinpointAppRead(d, meta)
+}
+
+func resourceAwsPinpointAppRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	log.Printf("[INFO] Reading Pinpoint App Attributes for %s", d.Id())
+
+	app, err := conn.GetApp(&pinpoint.GetAppInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	settings, err := conn.GetApplicationSettings(&pinpoint.GetApplicationSettingsInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	d.Set("name", app.ApplicationResponse.Name)
+	d.Set("application_id", app.ApplicationResponse.Id)
+
+	d.Set("campaign_hook", flattenCampaignHook(settings.ApplicationSettingsResource.CampaignHook))
+	d.Set("limits", flattenCampaignLimits(settings.ApplicationSettingsResource.Limits))
+	d.Set("quiet_time", flattenQuietTime(settings.ApplicationSettingsResource.QuietTime))
+
+	return nil
+}
+
+func resourceAwsPinpointAppDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	log.Printf("[DEBUG] Pinpoint Delete App: %s", d.Id())
+	_, err := conn.DeleteApp(&pinpoint.DeleteAppInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func expandCampaignHook(d *schema.ResourceData) *pinpoint.CampaignHook {
+	configs := d.Get("campaign_hook").(*schema.Set).List()
+	if configs == nil || len(configs) == 0 {
+		return nil
+	}
+
+	m := configs[0].(map[string]interface{})
+
+	ch := &pinpoint.CampaignHook{}
+
+	if m["lambda_function_name"] != nil {
+		ch.LambdaFunctionName = aws.String(m["lambda_function_name"].(string))
+	}
+
+	if m["mode"] != nil {
+		ch.Mode = aws.String(m["mode"].(string))
+	}
+
+	if m["web_url"] != nil {
+		ch.WebUrl = aws.String(m["web_url"].(string))
+	}
+
+	return ch
+}
+
+func flattenCampaignHook(ch *pinpoint.CampaignHook) []interface{} {
+	l := make([]interface{}, 0)
+
+	m := map[string]interface{}{}
+
+	if ch.LambdaFunctionName != nil {
```

Comment:
You can remove the `nil` checks by using the AWS Go SDK helpers here, e.g.

```go
m["lambda_function_name"] = aws.StringValue(ch.LambdaFunctionName)
```

Especially for `ch.Mode` below, which does not need the `!= ""` check.

Reply:
I've been struggling with this while testing the resource. The issue I've encountered is that `GetApplicationSettings` returns an empty object for `campaign_hook`, `limits` and `quiet_time` even if none of the children attributes is set. 

In this case, if I used 
```
m["lambda_function_name"] = aws.StringValue(ch.LambdaFunctionName)
m["mode"] = aws.StringValue(ch.Mode)
m["web_url"] = aws.StringValue(ch.WebUrl)
```
I'd get a map filled with empty values, but that would cause failures on the state diff (for example, on the `_basic` test):
```
        DIFF:

        UPDATE: aws_pinpoint_app.test_app
          campaign_hook.#: "1" => "0"

        STATE:

        aws_pinpoint_app.test_app:
          ID = ff3857dd45734e7ab07d9cb5141f0221
          provider = provider.aws
          application_id = ff3857dd45734e7ab07d9cb5141f0221
          campaign_hook.# = 1
          campaign_hook.0.lambda_function_name =
          campaign_hook.0.mode =
          campaign_hook.0.web_url =
          name = terraform-20180928131703355600000001
```

Any suggestions on how to handle this scenario?

Reply:
Great question! We'll be looking to come up with a better solution to this problem in the future, but for now, you can add the following `DiffSuppressFunc` to the `campaign_hook` attribute that ignores that the API is always returning the "container" attribute:

```go
DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
  if old == "1" && new == "0" {
    return true
  }
  return false
},
```

Reply:
Thanks! Totally missed `DiffSuppressFunc` while exploring the code looking for a workaround :D Fixed in https://github.com/terraform-providers/terraform-provider-aws/pull/5956/commits/4d398847e3e86408f66734b574bedfc086b6d493